### PR TITLE
Update to THREE r70 and some major fixes

### PR DESCRIPTION
**Pull request by psoetens**
This pull request is the result of 2 months coding against ros3djs, and fixing issues on the go. Finally, it also upgrades the codebase to THREE r70, which had quite some API changes since r63.

Apart from the modification of some default values, which could lead to discussion, I believe this patch series fixes a considerable number of issues in the ros3djs codebase, while maintaining compatibility.

For example, most objects now have management code to cleanly cleanup TFClient subscriptions if they're no longer needed, such that the tf_web_republisher is updated.

Also, zooming is now more responsive. I had to disable the highlighter rendering since this was not working well in r70. Someone should re-enable it and test it.

Finally, a new PointCloud object was added which is an alternative to DepthCloud when the cloudsize < 320x240, but this is still in development.

One incompatibility is that for DepthCloud.js, we changed the distance from 0-4m to 0-2m and use each RGB channels for depth values. The result is a 6x better resolution (2x3), but it requires an update to the depthcloud_encoder as well (pull request will be created). I however seriously doubt anyone was using this code (due to serious issues in the web_video_server mp4 streaming).

### CODE REVIEW

**Comment by rctoris** [comment]
Thanks!

Merge is broken after some other PR came in. Will have to do some good testing and running this given the scale of the PR.

**Comment by psoetens** [comment]
I pulled in the upstream changes. I kept everywhere the upstream parts, except for two call changes for setting the scale, by using .set() instead of individual .x/.y/.z setting or new Vector3D assignment (which doesn't work,: https://github.com/mrdoob/three.js/wiki/Migration#r67--r68 ).

**Comment by rctoris** [comment]
The removal of highlighting is actually more than a minor issue I think. For new users (not developers) the highlighting is essential for intuitive interaction why identifying objects they can interact with.

**Comment by rethink-rlinsalata** [comment]
Hi @psoetens - what versions of the dependency libraries are you using? I'm assuming three.js is r70, but what about the ColladaLoader's, etc? I tried to run your branch and ran into some trouble with undefined functions.

As a suggestion, it might be good to update the README.md file with the new version numbers you are using (though I guess the CDN will need to be updated later), since AFAIK, that's the only "manifest" in this repo that defines the version dependencies

 (-- that was actually another question I had for @rctoris , is that true? or is there another manifest somewhere so node or jenkins know what versions to use?)

**Comment by rctoris** [comment]
The READMEs list the tested versions of libs used that are known to work. Things like ColladaLoader are taken from the ThreeJS branch for a given version so an update to a ThreeJS version would mean needing to update to a newer version of ColladaLoader on the CDNs. 

I'm not sure what you mean by how node and jenkins know what to use?

**Comment by T045T** [comment]
+1 for updating the README so it points to the correct versions.

@psoetens, I'd like to take a look at highlighting - can you elaborate a bit on when and how it breaks using r70?

**Comment by rethink-rlinsalata** [comment]
Sorry, I realize that question didn't really make sense in this context.  I think I was thinking about unit tests, or if there was a way to install the package+dependencies with node / npm if it specified the dependencies to pull in (I don't really know node/npm that well yet, so this could very easily make no sense).

Is ColladaLoader2 now just a static snapshot ( I think it originally came from https://github.com/rmx/threejs-collada ?), or is the source code still being updated.  ColladaLoader2 seems broken with THREE r70; there have been a lot of changes in ColladaLoader from r60's to r70 of THREE, and some of the functions have gone away or changed.

I suppose I will/should convert over to STL's at some point, but I just want to make sure things are changing one step at a time.

**Comment by rethink-rlinsalata** [comment]
On another note, I'm still seeing the memory leak of tf callbacks in this branch #105.  It sounded like you were trying to address that in the tf cleanup you mentioned @psoetens, but maybe something was lost in the merges? Or was your cleanup addressing a different part?  If so, do you think there is a better way to do it, that is more in line with your updates?

**Comment by rctoris** [comment]
ColladaLoader2 was from a third party developer who created the package. It unfortunately hasn't been updated as you can see from the GitHub page you listed. ColladaLoader is a snapshot from the ThreeJS repo (https://github.com/mrdoob/three.js/blob/2d59713328c421c3edfc3feda1b116af13140b94/examples/js/loaders/ColladaLoader.js).

**Comment by psoetens** [comment]
Sorry, I missed this discussion... Some preliminary answers:

- I tested against r70 and for the rest the 'existing' versions. So only three.js was upgraded in my tests.
- I did a necessary fix in the highlighter, which you can see in the commit log. But in a commit some time later, I disabled it (by commenting it) while tracing an unrelated bug. I didn't turn it on anymore since I couldn't test it, but it could be fixed actually. Just needs testing/confirming
- About the tf leaks: my patches require that you 'manually' remove the tf subscriptions of a sceneNode by calling node.removeTF() . I have no idea on how to implement a proper 'destructor' in javascript, such that the cleanup is automatically.
- mac vs that :  sorry, I can update that too for style correctness
- PointCloud.js is beta... also adds an extra dependency if you want to use it, which is not documented yet. I forgot to mention that.

In summary, I agree that the pull request is not yet up to level, but I didn't want to delay access to some of the fixes I did for the rest of the community...

**Comment by DLu** [comment]
Is there a timeframe for upgrading to THREE r70? I have no idea how much it would break, but I am working on my own PointCloud object, and the switch would break if r70 is used.

**Comment by rctoris** [comment]
No timeframe at this point unless there is some need for it in my opinion. If there is an update, we would need to update many other files as well.

**Comment by rctoris** [comment]
Closing due to inability to merge/no recent activity. Please re-open a bug ticket for discussion if still needed.
